cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/schedule/ProgramSchedule.java

```diff
@@ -43,17 +45,19 @@ public class ProgramSchedule {
   private final long timeoutMillis;
 
   public ProgramSchedule(String name, String description,
-                         ProgramId programId, Map<String, String> properties,
+                         ProgramReference programReference, Map<String, String> properties,
                          Trigger trigger, List<? extends Constraint> constraints) {
-    this(name, description, programId, properties, trigger, constraints, Schedulers.JOB_QUEUE_TIMEOUT_MILLIS);
+    this(name, description, programReference, properties, trigger, constraints, Schedulers.JOB_QUEUE_TIMEOUT_MILLIS);
   }
 
-  public ProgramSchedule(String name, String description,
-                         ProgramId programId, Map<String, String> properties,
-                         Trigger trigger, List<? extends Constraint> constraints, long timeoutMillis) {
+  public ProgramSchedule(String name, String description, ProgramReference programReference,
+                         Map<String, String> properties, Trigger trigger, List<? extends Constraint> constraints,
+                         long timeoutMillis) {
     this.description = description;
-    this.programId = programId;
-    this.scheduleId = programId.getParent().schedule(name);
+    // CDAP-19989: this attribute should be programRef instead of programId, keep it as is by adding the default version
```

Comment:
the existing ProgramSchedule is serialized with ProgramId (program:ns.app.version.programType.program), if we change it to ProgramReference then it will have some trouble deserializing old ProgramSchedules. and thus not back compat.

related code:
https://github.com/cdapio/cdap/blob/develop/cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/schedule/store/ProgramScheduleStoreDataset.java#L392

Reply:
Are we getting an error when we try to deserialize? We should be able to deserialize a ProgramId into a ProgramReference since the fields are all named the same, unless we're doing something strange in the gson registered type adapters

Reply:
No error during deserializing, it will construct a ProgramReference object, but incorrectly tho because EntityType would still remain "program" (eg. programRef.toString() will be program:ns.app.programType.program rather than programReference:ns.app.programType.program).  So if we change it from programId to programRef,  programRef1.equals(programRef2) would not work properly.

Reply:
ah I see. Do you know why we needed to add those reference entity types? In other words, why couldn't a ProgramReference be of type Program?

I'm concerned that we might unknowingly have this problem in other places where a reference replaced an id.

Reply:
yeah, making ProgramReference & ApplicationReference be of type Program & Application makes more sense to me, and will avoid situations like this. 
FYI @de-lan, I can go ahead give that a try, let me know if anything I should be aware for this change.

Reply:
are the reference entity types there in 6.8.0? If so that would cause complications with some data already stored with that entity type.

Reply:
ok it looks like they are there in 6.8.0. If there wasn't any Reference class being serialized and stored anywhere, it is ok to change remove those entity types. If they are getting stored then we'll need some other plan.

Reply:
it looks like the EntityType exists because there needs to be a mapping from type to class in order to deserialize entities using the EntityIdTypeAdapter. So it wouldn't make sense to map the same type to multiple classes, we can just keep what you have for now as you've already opened a jira to improve things.